Added code to support IAM API Key Authentication

athira-varghese opened the pull request:
https://github.ibm.com/Observability/dragonlog/issues/1352

athira-varghese committed:
Develop code to support IAM API Key Authentication

athira-varghese committed:
Added code to support IAM API Key Authentication

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/cos_object_client.go:
```diff
@@ -135,6 +152,29 @@ func validate(cfg COSConfig) error {
 		return errEmptyEndpoint
 	}
+
+	if cfg.ApiKey.String() == "" {
```
If we don't have complete HMAC credentials or API Key we should raise an error that there must be at least one authentication method enabled.

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/cos_object_client.go:
```diff
@@ -135,6 +152,29 @@ func validate(cfg COSConfig) error {
+		return errEmptyApiKey
+	}
+
+	if cfg.AuthEndpoint == "" {
```
This check is needed only if we use API Key authentication

tareqmamari commented on pkg/storage/chunk/client/ibmcloud/cos_object_client.go:
```diff
@@ -135,6 +152,29 @@ func validate(cfg COSConfig) error {
+		return errAuthEndpoint
+	}
+
+	if cfg.ServiceInstanceID == "" {
```
this one as well, we only need it if we use api key.

athira-varghese committed:
Update the conditions for api key

tareqmamari committed:
use default auth endpoint for cos

Use default auth endpoint instead of forcing to pass auth endpoint always
for COS storage provider.

shahulsonhal commented on pkg/storage/chunk/client/ibmcloud/cos_object_client.go:
```diff
@@ -26,13 +28,16 @@ import (
 	"github.com/weaveworks/common/instrument"
 )
 
+const DEFAULT_COS_AUTH_ENDPOINT = "https://iam.cloud.ibm.com/identity/token"
```
Is there any specific reason to capitalise the constant? Can't we follow the normal go variable convention here? Since in Go the things are exported if it's start with a capital letter. 
```suggestion
const defaultCOSAuthEndpoint = "https://iam.cloud.ibm.com/identity/token"
```

adityacs replied:
The general convention is to capitalize the constants. Agree on the const being exported but in this case don't have a strong argument whether it should be exported or not.

tareqmamari replied:
Let's be consistent with what is already in Loki.

shahulsonhal committed:
Replace const to defaultCOSAuthEndpoint

Replace DEFAULT_COS_AUTH_ENDPOINT with defaultCOSAuthEndpoint

Also fixed error fmt.Errorf call needs 6 args but has 7 args

amankrsingh2000 committed:
Added new cases in Test_COSConfig

shahulsonhal reviewed (changes requested):
@amankrsingh2000 It seems like there is an indentation (formatting) issue in your last commit. It should be auto formatted when you save the file (it wont work if you have turned on auto saving in your editor). Could you please check?

athira-varghese committed:
Added unit test for IAM API Key Authentication

tareqmamari committed:
iam api key auth test formatting

These are just minor formatting changes for IAM API key authentication
tests.

tareqmamari commented:
I am merging this PR, if any further changes are needed, please feel free to open another one.
